### PR TITLE
Resolves 1892: compute view factor for rht

### LINKED ISSUE #1892: Compute ray tracing view factor for RHT
add view factor computation 

it uses nanoflann #1893 
see also #1806 

/cc @alixkien

### LINKED ISSUE #1893: add nanoflann to feelpp
nanoflann is an efficient modern c++ implementation of kdtree data structure

it should go in `feelpp/feel/feelcore`

### PR DESCRIPTION
**Pull request by lberti**
- View factor computation via raytracing is performed.
- Two test cases on a cubic and cylindrical cavity are proposed: checks on view factors summing to 1 and comparison with analytical solutions are performed
- Test cases for view factors with quadrature rules have been updated and separated from raytracing ones.
- A Bounding Volume Hierarchy structure is used, and an elementary stackless raytracing algorithm is implemented, as proposed in 
> Hapala, Michal, et al. "Efficient stack-less bvh traversal for ray tracing." Proceedings of the 27th Spring Conference on Computer Graphics. 2011.

- closes #1892 
- closes #1893

### CODE REVIEW

File: feelpp/feel/feelmesh/bvh.hpp [hunk 1]
```diff
@@ -0,0 +1,560 @@
...
+#include <boost/tuple/tuple_io.hpp>
+
+#include <feel/feelcore/feel.hpp>
+#include <feel/feelalg/glas.hpp>
+
+#include <feel/feel.hpp>
+
+namespace Feel
+{
+    class Ray_bvh
```
Comment (prudhomm) [inline feelpp/feel/feelmesh/bvh.hpp]: please rename to `BVHRay`
see naming conventions here https://docs.feelpp.org/dev/latest/reference/codingrules.html#_naming_convention

File: feelpp/feel/feelmesh/bvh.hpp [hunk 2]
```diff
@@ -0,0 +1,560 @@
...
+    };
+
+    inline constexpr float robust_const(int n) 
+    {
+        double epsil = std::numeric_limits<double>::epsilon() * 0.5;
+        return (n * epsil) / (1 - n * epsil);
+    }
+
+    template<int nDim>
+    class BVH_tree
```
Comment (prudhomm) [inline feelpp/feel/feelmesh/bvh.hpp]: idem `BHVTree`, and use `BHVRay` to use the same naming

File: feelpp/feel/feelmesh/bvh.hpp [hunk 3]
```diff
@@ -0,0 +1,560 @@
...
+                    M_bound_min[k] = M_bound_min[k] - 2*FLT_MIN;
+                    M_bound_max[k] = M_bound_max[k] + 2*FLT_MIN;
+                }                
+                BVHPrimitiveInfo primitive_i(e.id(),M_bound_min,M_bound_max);
+                M_primitiveInfo.push_back(primitive_i);
+                index_Primitive++;
+            }
+        }    
+
+        struct BVH_node
```
Comment (prudhomm) [inline feelpp/feel/feelmesh/bvh.hpp]: `BHVNode`

File: feelpp/feel/feelmesh/bvh.hpp [hunk 4]
```diff
@@ -0,0 +1,560 @@
...
+                parent = current_parent;
+                M_bounds_min = new_Bounds_min(child0->M_bounds_min,child1->M_bounds_min);
+                M_bounds_max = new_Bounds_max(child0->M_bounds_max,child1->M_bounds_max);
+                M_centroid = (M_bounds_min + M_bounds_max) *0.5;
+                splitaxis = splitaxisIn;
+                nPrimitives = 0;
+                LOG(INFO) <<fmt::format("Internal node built: M_bounds_min {}, M_bounds_max {}, splitaxis {}",M_bounds_min,M_bounds_max,splitaxis);
+            }
+
+            Eigen::VectorXd new_Bounds_min(Eigen::VectorXd bounds1,Eigen::VectorXd bounds2)
```
Comment (prudhomm) [inline feelpp/feel/feelmesh/bvh.hpp]: newBoundsMin

File: feelpp/feel/feelmesh/bvh.hpp [hunk 5]
```diff
@@ -0,0 +1,560 @@
...
+            {
+                Eigen::VectorXd newBoundsMin(bounds1.size());
+                for(int i=0;i<bounds1.size();i++)
+                    newBoundsMin[i] = std::min(bounds1[i],bounds2[i]);
+                return newBoundsMin;
+            }
+            Eigen::VectorXd new_Bounds_max(Eigen::VectorXd bounds1,Eigen::VectorXd bounds2)
```
Comment (prudhomm) [inline feelpp/feel/feelmesh/bvh.hpp]: `newBoundsMax`

File: feelpp/feel/feelmesh/bvh.hpp [hunk 6]
```diff
@@ -0,0 +1,560 @@
...
+                    if (t2 > tmax)
+                        tmax = t2;
+                    if (tmin > tmax)
+                        return false;
+                }
+                
+                return true;         
+            }
+            
+            bool check_intersection_with_segment(Ray_bvh const& ray)
```
Comment (prudhomm) [inline feelpp/feel/feelmesh/bvh.hpp]: `checkIntersectionWithSegment`

File: feelpp/feel/feelmesh/bvh.hpp [hunk 7]
```diff
@@ -0,0 +1,560 @@
...
+                    w_[0] =  ray.origin[0] + ray.dir[0]*t1;
+                    w_[1] =  ray.origin[1] + ray.dir[1]*t1;
+                    return true;
+                }
+                
+                return false;
+                
+            }
+
+            std::pair<bool,double> check_intersection_with_segment(matrix_node_type const& nodes, Ray_bvh const& ray)
```
Comment (prudhomm) [inline feelpp/feel/feelmesh/bvh.hpp]: idem use camel case naming

File: feelpp/feel/feelmesh/bvh.hpp [hunk 8]
```diff
@@ -0,0 +1,560 @@
...
+                    w_[1] =  ray.origin[1] + ray.dir[1]*t1;
+                    return std::make_pair(true,t1);
+                }
+
+                return std::make_pair(false,t1);            
+                
+            }
+            
+            // Verify if the ray intersects the element
+            std::pair<bool,double> check_intersection_with_triangle( Ray_bvh const& ray)
```
Comment (prudhomm) [inline feelpp/feel/feelmesh/bvh.hpp]: camel case

File: feelpp/feel/feelmesh/bvh.hpp [hunk 9]
```diff
@@ -0,0 +1,560 @@
...
+                                    recursiveBuild( node, (cut_dimension+1)%nDim, mid, end_index_primitive, orderedPrims));
+            }
+
+            return node;
+        }
+
+        // Verify if the ray intersects the whole bounding structure
+        // Returns the integer corresponding to the intersected element
+        // If no element is intersected, return -1
+        int ray_search( Ray_bvh const& rayon,std::string s)
```
Comment (prudhomm) [inline feelpp/feel/feelmesh/bvh.hpp]: camel case

File: feelpp/feel/feelmesh/bvh.hpp [hunk 10]
```diff
@@ -0,0 +1,560 @@
...
+                    }
+                    else
+                    {                        
+                        current_node = current_node->nearChild(rayon);
+                        state='P'; // the current node has been accessed from its parent                     
+                    }
+                    break;
+
+                default:
+                    std::cout << "ERROR: None of the previous cases has been traversed" <<std::endl;
```
Comment (prudhomm) [inline feelpp/feel/feelmesh/bvh.hpp]: use `LOG(ERROR)`

Comment (prudhomm) [reply feelpp/feel/feelmesh/bvh.hpp]: eventually throw an exception

File: feelpp/feel/feelviewfactor/raytracingviewfactor.hpp [hunk 11]
```diff
@@ -13,36 +13,60 @@
...
+    
 // Compute random direction, uniformly distributed on the sphere or on a circle
-std::vector<double> get_random_direction(std::vector<double> &random_direction)
-{
-    unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
-    unsigned seed1 = std::chrono::system_clock::now().time_since_epoch().count();              
-
-    std::default_random_engine generator1(seed);
-    std::default_random_engine generator2(seed1);
+void get_random_direction(std::vector<double> &random_direction, std::mt19937 & M_gen, std::mt19937 & M_gen2,Eigen::VectorXd normal)
```
Comment (prudhomm) [inline feelpp/feel/feelviewfactor/raytracingviewfactor.hpp]: camel case

File: feelpp/feel/feelviewfactor/raytracingviewfactor.hpp [hunk 12]
```diff
@@ -13,36 +13,60 @@
...
+            matrix1.col(2) = crossProd1;
+            
+            matrix2.row(0) = normal;
+            matrix2.row(1) = crossProd3;
+            matrix2.row(2) = crossProd1;
+            
+            
+            direction = matrix1 * (matrix2 * direction);
+
+            random_direction[0]=direction[0];
```
Comment (prudhomm) [inline feelpp/feel/feelviewfactor/raytracingviewfactor.hpp]: these element wise operation are inefficient, write `random_direction=direction`

File: feelpp/feel/feelviewfactor/raytracingviewfactor.hpp [hunk 13]
```diff
@@ -97,7 +121,7 @@ bool isOnSurface(Eigen::VectorXd const &point,Eigen::VectorXd const &el_p1,Eigen
     auto c = (el_p1+el_p2+el_p3)/3.;
     auto elem_area = element_area(c, el_p1,el_p2,el_p3); 
     auto area = element_area(point, el_p1,el_p2,el_p3);
```
Comment (prudhomm) [inline feelpp/feel/feelviewfactor/raytracingviewfactor.hpp]: camel case

File: feelpp/feel/feelviewfactor/raytracingviewfactor.hpp [hunk 14]
```diff
@@ -170,66 +227,21 @@ class RayTracingViewFactor : public ViewFactorBase<MeshType>
     std::vector<std::string> M_markers_string,list_of_bdys;
     std::vector<int> M_markers_int;
     int M_Nrays;
+    int M_Nthreads;
     std::vector<int> M_point_indices;
-    KdTree_type* M_kdtree_boundariesptr;
     Eigen::VectorXd M_view_factor_row;
-    bool M_inward;
-    KDTree M_kd_tree_feelpp;
+    BVH_tree<MeshType::nDim> M_bvh_tree;
```
Comment (prudhomm) [inline feelpp/feel/feelviewfactor/raytracingviewfactor.hpp]: BHVTree

File: testsuite/feelmesh/CMakeLists.txt [hunk 15]
```diff
@@ -25,3 +25,6 @@ if ( FEELPP_HAS_PARMMG OR FEELPP_HAS_MMG )
                        test_remesh.json test_remesh_relations.json test_remesh_curved.json 
                    TIMEOUT 500)
 endif()
+
+feelpp_add_application(bvh2D SRCS test_bvh.cpp DEFS FEELPP_DIM=2 CFG test_bvh2d.cfg GEO domain_2d.geo )
+feelpp_add_application(bvh3D SRCS test_bvh.cpp DEFS FEELPP_DIM=3 CFG test_bvh3d.cfg GEO cubic_cavity.geo )
```
Comment (prudhomm) [inline testsuite/feelmesh/CMakeLists.txt]: probably not good, it does not follow the test naming convention : `feelpp_test_<name of the test>` either use `feelpp_add_test` or use `feelpp_add_application(test_bvh2D ...`

File: testsuite/feelmesh/test_bvh.cpp [hunk 16]
```diff
@@ -0,0 +1,69 @@
+// #define BOOST_TEST_MODULE bvh tests
+// #include <feel/feelcore/testsuite.hpp>
+
+
+#include <feel/feelmesh/bvh.hpp>
+using namespace Feel;
+
+int main(int argc , char **argv)
```
Comment (prudhomm) [inline testsuite/feelmesh/test_bvh.cpp]: there are no test here, please use BOOST_TEST

File: testsuite/feelviewfactor/test_viewfactor_quadrature.cpp [hunk 17]
```diff
@@ -0,0 +1,573 @@
...
+//     std::ifstream f(  Environment::expand(jsons[0]) );
+//     f >> j;
+//     BOOST_TEST_MESSAGE(fmt::format("j: {}",j.dump(1)));
+ 
+//     RayTracingViewFactor<MeshType> rtvf( mesh, j );
+    
+//     rtvf.compute();
+//     BOOST_TEST_MESSAGE( fmt::format("Max dev reciprocity {}", rtvf.maxDevReciprocity()));
+    
+//     BOOST_TEST_MESSAGE( fmt::format("{}", rtvf.viewFactors() ) );
```
Comment (prudhomm) [inline testsuite/feelviewfactor/test_viewfactor_quadrature.cpp]: perhaps remove all this commented code, it should be cleaned up

**Review by prudhomm — CHANGES_REQUESTED** [review]

**Comment by prudhomm** [comment]
@lberti I merged the PR #2039 so there will be conflicts with with this one. Please do not remove the work I did on nanonflann it will certainly be useful in other context.

**Comment by lberti** [comment]
@prudhomm I have made the required changes and the test suite is completed successfully

**Review by prudhomm — APPROVED** [review]

**Pull request merged**